### ISSUE TITLE
Error: Setting a field value requires a model to be set

### ISSUE THREAD

**Issue by Slawikus** [adamghill/django-unicorn#168]
Hi there,

thanks a lot for creating for us a new magic!

I am trying to get through the docs and being stuck on DbModel part: https://www.django-unicorn.com/docs/django-models/#dbmodel
I've followed the examples, here is what I have:

model:
```
class Todo(models.Model):
    description = models.CharField(max_length=50)
    is_completed = models.BooleanField(default=False, blank=True)
    due_date = models.DateField(null=True, blank=True)
```

unicorn view ('add_todo_test.py')
```
from django_unicorn.components import UnicornView
from django_unicorn.db import DbModel

from todos.models import Todo


class AddTodoTestView(UnicornView):
    class Meta:
        db_models = [DbModel("todo", Todo)]

    def save(self):
        print("A new book will be created automatically")
        pass

```

and component ('add-todo-test.html')
```
<div>
  <div>
    <input unicorn:db="todo" unicorn:field.defer="description" type="text" id="description" />
    {{ todo.description }}
  </div>
  <div>
    <input unicorn:db="todo" unicorn:field.defer="due_date" type="text" id="due_date" />
    {{ todo.due_date }}
  </div>
  <button unicorn:click="save">Save</button>
</div>
```

Component is rendered, but it doesn't work and I get the following error message in browser's dev console:
```
component.js:404 Uncaught Error: Setting a field value requires a model to be set
    at component.js:404
    at Array.forEach (<anonymous>)
    at Component.setDbModelValues (component.js:398)
    at Module.componentInit (unicorn.js:35)
    at (index):92
(anonymous) @ component.js:404
setDbModelValues @ component.js:398
componentInit @ unicorn.js:35
(anonymous) @ (index):92
```

I simply cannot understand what am I doing wrong  :)

P.S. Don't know if that plays any role, but in PyCharm I get the following warning:
![Screenshot 2021-03-12 at 19 55 42](https://user-images.githubusercontent.com/7048942/110985598-ef558d00-836c-11eb-967a-22c508fc5bcf.png)

**Comment by adamghill** [comment]
Thanks for the detailed example! I created a branch with the code to replicate and I think if you change the component to something like this, it _should_ work. 

```
<div unicorn:db="todo">
  <div>
    <input unicorn:field.defer="description" type="text" id="description" />
    {{ todo.description }}
  </div>
  <div>
    <input unicorn:field.defer="due_date" type="text" id="due_date" />
    {{ todo.due_date }}
  </div>
  <button unicorn:click="save">Save</button>
</div>
```
You can see the branch where I was testing this here: https://github.com/adamghill/django-unicorn/commit/3f2ed39fefda91830e1ebc9f3a82018fc5173f49#diff-e3803954fdf2aed5407a8d35f9e8e2a0f76058cebe6d276b39f7682d5ec7427eR1-R11.

If that functionality works, I will update the documentation for now, and then also create an issue to support `unicorn:db` inside the same element as `unicorn:field`. Let me know!

**Comment by Slawikus** [comment]
As you have suggested I've moved the `unicorn:db` declaration from the inputs into the parent div, but I get the same error message is JS console, component as before is rendered but doesn't work (if one types something in the inputs it is not reflected on the page).

**Comment by adamghill** [comment]
@Slawikus if you are able to test https://github.com/adamghill/django-unicorn/pull/173 that would be great!

**Comment by adamghill** [comment]
https://www.django-unicorn.com/docs/changelog/#v0-25-0 has been released with this fix! 🎉

**Comment by Slawikus** [comment]
Yeah, now it works, one doesn't get any error message in JS console - thanks a lot, Adam!

Few things I have noticed:

1) if one removes ".defer" modifier - new model is created on each change of the field, is this intended behaviour?

2) {{ todo.description }} always stays empty, it doesn't show anything, even if "defer" is removed from the field definition.

**Comment by adamghill** [comment]
>is this intended behaviour?

It was, but I think it might be confusing? https://github.com/adamghill/django-unicorn/blob/master/django_unicorn/views.py#L412-L417 is where I check for fields and either update the model or save it. I wonder if it makes more sense to just set the model's fields in memory and let the user update/save it explicitly in a method? What do you think?

>{{ todo.description }} always stays empty

I'll take a look at why that might be happening next.

**Comment by Slawikus** [comment]
In fact for me this behavour was confusing. I somehow expected that any changed to the models must be explicitly provided in the code. For example in the documentation it was written that irrelevant of which method one calls for the mode instance it will be saved. Wouldn't it be better to provide some hooks in the base view which user can override to implement the behaviour he needs? Like save update delete methods? So yes, I think explicitly it would be the best.

**Comment by adamghill** [comment]
You are definitely right.

I'm re-thinking how the database model integration works completely because 1) having a separate "API" (e.g. `unicorn:db`, `unicorn:pk`, `unicorn:field`) makes things really confusing, and 2) there is too much magic that happens for database models currently and everything should be more explicit. I'm also thinking I can get rid of the `db_model` decorator and instead use type hints which would be cleaner.

I'm about half-way through improving the integration with database models, but just wanted to say thanks for the help and pushing to make the API better. To that end, my plan is to leave this current implementation for the next version, but mark it as deprecated and then remove it in the version after.

I'll attach a PR with the initial code for these changes soon.

**Comment by adamghill** [comment]
PR https://github.com/adamghill/django-unicorn/pull/177 provides the pieces so that Django models and querysets can "just work" similarly to everything else. I think it'll be easier for users to use and I'll also be able to remove some client-side code that handled the complexity, so it seems like a good move going forward. I'll attach a PR for the documentation changes as well so you can see what would change for developers before merging this branch into master.

**Comment by adamghill** [comment]
#177 has been merged into master.

Documentation changes that detail how to interact with Django models in the upcoming 0.26.0: https://github.com/adamghill/django-unicorn.com/pull/14. Lots of back-end changes here, but hopefully it contains less magic and is more explicit. Let me know what you think!

**Comment by adamghill** [comment]
Just a follow-up that those changes for using Django models have been released in 0.26.0. Documentation is at: https://www.django-unicorn.com/docs/django-models/.
